<h1>Základní Datové typy a Operátory</h1>
<h2>Základní Datové typy</h2>
<p>Každá proměnná má svůj datový typ, a to podle toho, jakou do ní uložíme hodnotu.<br>Python má takzvané dynamické datové typování - tedy datový typ proměnné se může v průběhu programu měnit - záleží, jakou hodnotu jsme do proměnné uložili naposledy.</p>
<table>
<tr> <th>Význam</th> <th>Anglicky</th> <th>Označení v Pythonu</th> <th>Možné hodnoty</th> </tr>
<tr> <td>Celé číslo</td> <td>integer</td> <td>int</td> <td> 0 , 178 , -9 </td> </tr>
<tr><td>Desetinné číslo</td><td>float</td><td>float</td> <td> 0.0 , 3.14 , -2.71  </td> </tr>
<tr><td>Text (řetězec znaků)</td><td>string</td><td>str</td> <td> "A" , "Ahoj" , "Jarda" </td> </tr>
<tr><td>Pravda/Nepravda</td><td>boolean</td><td>bool</td> <td>True , False </td> </tr>
</table>

<h2>Funkce <code>type()</code></h2>
<p>Funkce <code>type()</code> nám vrátí datový typ proměnné kterou dáme do jejího argumentu.<br>
Pro vypsání výsledku funkce do konzole, použijeme funkci <code>print()</code></p>

In [ ]:
number = 34
name = "Honza"
boolean_value = False
another_number = 56.981

print(number, type(number))
print(name, type(name))
print(boolean_value, type(boolean_value))
print(another_number, type(another_number))

34 <class 'int'>
Honza <class 'str'>
False <class 'bool'>
56.981 <class 'float'>


Co znamená slovo `class` zjistíme později v materiálu k objektovému programování.

<h2>Přetypování</h2>
<p>Pomocí funkcionality přetypování lze datový typ proměnné přetypovat na jiný datový typ.</p>

In [ ]:
promenna = 64 # int
promenna2 = float(promenna)
print(promenna, type(promenna), promenna2, type(promenna2))

cislo = 12
retezec = str(cislo)
zpet_cislo = int(retezec)
print(cislo, type(cislo), retezec, type(retezec), zpet_cislo, type(zpet_cislo))

64 <class 'int'> 64.0 <class 'float'>
12 <class 'int'> 12 <class 'str'> 12 <class 'int'>


Přetypování často používáme při načítání vstupu od uživatele. Funkce <code>input()</code> vrací vždy string a pokud chceme s hodnotou pracovat jako s číslem, musíme ji přetypovat.

In [ ]:
my_integer = int(input("Zadejte cislo:"))
print("Zadal jsi:", my_integer, type(my_integer))

<h2>Operátory se základními datovými typy</h2>
<p>Jak název napovídá, operátor je znak, který provede nějakou operaci.<br>Při použití operátorů je důležité s jakým <strong>datovým typem</strong> jej používáme - výsledek stejného operátoru na různém datovém typu se obykle liší a někdy ani nelze operátor použít.</p>

<h3>=</h3>
<p>Operátor <code>=</code> znamená "ulož".<br>Vezme tedy <strong>hodnotu vpravo</strong> a uloží ji do <strong>proměnné vlevo</strong>.<br>Pokud bude vpravo proměnná, použije se hodnota této proměnné, ne proměnná samotná.</p>
<p>Srovnejte s použitím znaménka = v matematice. V čem se liší?</p>

<h3>+ -</h3>
<p>Operátory <code>+</code> a <code>-</code> provádí s čísly očekávané sčítání a odčítání.<br>Co však znamená sečíst/odečíst dva řetězce? Vyzkoušejte v kódu.</p>

In [1]:
a = 15
b = 37
scitani = a + b
odcitani = a - b
print("Cisla:", a, "+", b, "=", scitani, "|", a, "-", b, "=", odcitani)

napis1 = "Ahoj"
napis2 = "Jarda"
scitani = napis1 + napis2
print("Nápisy:",napis1,"+",napis2,"=",scitani)

odcitani = napis1 - napis2

Cisla: 15 + 37 = 52 | 15 - 37 = -22
Nápisy: Ahoj + Jarda = AhojJarda


TypeError: unsupported operand type(s) for -: 'str' and 'str'

<h3>*</h3>
<p>Operátor <code>*</code> provádí s čísly násobení.<br>Podobně jako v předchozím případě vynásobit dva řetězce nelze. Co nám však Python umožňuje, je vynásobit číslo a řetězec.</p>

In [2]:
a = 15
b = 37
nasobeni = a * b
print("Cisla:", a, "+", b, "=", nasobeni)

napis = "Ahoj"
nasobeni = a * napis
print("Číslo a nápis:",a,"*",napis,"=",nasobeni)

Cisla: 15 + 37 = 555
Číslo a nápis: 15 * Ahoj = AhojAhojAhojAhojAhojAhojAhojAhojAhojAhojAhojAhojAhojAhojAhoj


<h3>/</h3>
<p>Operátor <code>/</code> provádí s čísly dělení.<br>Jeho výsledkem je vždy <strong>float</strong>, bez ohledu na typ vstupních čísel a hodnotu výsledku.</p>

In [3]:
a = 15
b = 3
deleni = a / b
print("Integer:", a, "/", b, "=", deleni, type(deleni))

a = 15.5
b = 3.1
deleni = a / b
print("Float:", a, "/", b, "=", deleni, type(deleni))

Integer: 15 / 3 = 5.0 <class 'float'>
Float: 15.5 / 3.1 = 5.0 <class 'float'>


<h3>// %</h3>
<p>Operátory <code>//</code> a <code>%</code> znáte z 2. třídy základní školy.<br>Předtím, než jste se dozvěděli o desetinných čísel jste dělili dvě celá čísla na "výsledek" a "zbytek". "Výsledek" je ve skutečnosti výsledek celočíselného dělení, které v Pythonu realizuje operátor <code>//</code> a "zbytku", který získáme pomocí <code>%</code> někdy říkáme <strong>modulo</strong>.</p>
<p>Možná pro vás trochu překvapivě můžeme "celočíselně se zbytkem" vydělit i datový typ float - v takovém případě bude výsledek také datového typu float.<br>Operátor <code>//</code> si pak můžete představit jako otázku "kolikrát se druhé číslo vejde do prvního?"<br>Operátor <code>%</code> jako otázku "co mi zbyde z prvního čísla, pokud odečtu druhé číslo tolikrát, dokud to půjde (ne do záporných čísel)?"</p>
<p>Proč je výsledek příkladu níže "divný" vám necháme jako procvičení hledání odpovědí na internetu.<br>Malá nápověda:</p><img src="https://www.meme-arsenal.com/memes/6f5ad8c64dac8a42cfc9960c5cbfe4a9.jpg" width="30%">

In [4]:
a = 15
b = 3
celociselne = a // b
zbytek = a % b
print("Integer:", a, "//", b, "=", celociselne, "zbytek", zbytek, type(celociselne) ,type(zbytek) )

a = 15.5
b = 3.1
celociselne = a // b
zbytek = a % b
print("Float:", a, "//", b, "=", celociselne, "zbytek", zbytek, type(celociselne) ,type(zbytek) )

Integer: 15 // 3 = 5 zbytek 0 <class 'int'> <class 'int'>
Float: 15.5 // 3.1 = 4.0 zbytek 3.0999999999999996 <class 'float'> <class 'float'>


<h3>**</h3>
<p>Operátor <code>**</code> provádí s čísly mocnění.<br>Datový typ výsledku odpovídá datovému typu vstupu.</p>
<p>Kontrolní otázka: Co znamená např.: $ 3^{0.5} $ ?</p>

In [5]:
a = 15
b = 3
mocnina = a ** b
print("Integer:", a, " na ", b, "=", mocnina, type(mocnina))

a = 15.5
b = 3.1
mocnina = a ** b
print("Integer:", a, " na ", b, "=", mocnina, type(mocnina))

Integer: 15  na  3 = 3375 <class 'int'>
Integer: 15.5  na  3.1 = 4898.106957108926 <class 'float'>


<p>Operátory mohou pracovat i přímo s hodnotami (hodnoty nemusí být uložené v proměnných).</p>

In [3]:
a = 5
nasobeni = a * 10
print("Desetinásobek",a,"je",nasobeni)

Desetinásobek 5 je 50


<h3>Operátory porovnání</h3>
<p>Tyto operátory porovnávají dvě hodnoty.<br>Jejich výsledkem je boolean (True/False)</p>
<h3>==</h3>
<p>Operátor <code>==</code> zjišťuje, zda jsou dvě hodnoty stejné</p>

In [6]:
print(2 == 5)
print(True == False)
print("ahoj" == "ahoj")
print("jedna" == "dva")

False
False
True
False


<h3>< > <= >=</h3>
<p>Operátory <code><</code>, <code>></code>, <code><=</code>, <code>>=</code> zjišťují, zda je první hodnota - menší - větší - menší nebo rovna - větší nebo rovna - druhé hodnotě</p>

In [7]:
print(2 < 5)
print(2 >= 2)
print("ahoj" < "ahoj")
print("jedna" < "dva") # proč je výsledek False?

True
True
False
False


<h3>Zkrácený zápis</h3>
<p>Operátory lze kombinovat tak, že nám usnadní psaní.</p>
<p>Např. pokud chceme přičíst hodnotu do proměnné, zápišeme příkaz následovně:</p>

In [8]:
var = 0
print(var)
var = var + 2
print(var)

0
2


Zápis jde však zjednodušit:

In [9]:
print(var)
var += 2 # přečte hodnotu proměnné var, přičte k hodnotě 2 a výsledek uloží zpět do proměnné var
print(var)

2
4


Tento zápis jde použít pro všechny matematické operace:

In [10]:
print(var)
var -= 4 # odečte od proměnné var 4
print(var)
var *= 2 # vynásobí proměnnou dvěma
print(var)
var /= 2 # vydělí proměnnou dvěma
print(var)
var %= 2 # uloží zbytek po dělení dvěma
print(var)

4
0
0
0.0
0.0


<h2>Cvičení</h2>
<h3>1</h3>
Napiš program, který počítá s dvěma desetinnými čísly.
<p>Sečte je, jejich součet umocní na třetí a poté vydělí číslem $\pi$ (3.14)<br>
Na nový řádek vypíše vstupní čísla, výsledek a datový typ výsledku.</p>

In [17]:
x = 4
y = 3
z= x + y
z= z**3
z= z/3.14
print(x,y,z,type(z))

4 3 109.23566878980891 <class 'float'>


<p>Výsledky pro Vaší kontrolu - referenční řešení má 4 řádky</p>
<table>
<tr>

<td><pre>
4 3 109.23566878980891 &ltclass 'float'&gt
</pre></td>
<td><pre>
1 1 2.5477707006369426 &ltclass 'float'&gt
</pre></td>
<td><pre>
10 21 9487.579617834395 &ltclass 'float'&gt
</pre></td>

</tr>
</table>

<h3>2</h3>
<p>Načtěte od uživatele celé číslo (užijte přetypování) a vynásobte ho dvěma.<br>Načtěte druhé celé číslo, celočíselně ho vydělte dvěma.<br>
Vypište obě zadaná čísla, na další řádek výsledky výpočtů a na třetí řádek zda se tyto dva výsledky rovnají.</p>

In [ ]:
a = int(input("zadejte první číslo"))


0 0
0 0
True


<p>Výsledky pro Vaší kontrolu - referenční řešení má 3 řádky</p>
<table>
<tr>

<td><pre>
Zadej první číslo: 5
Zadej druhé číslo: 20
5 20
10 10
True
</pre></td>
<td><pre>
Zadej první číslo: 8
Zadej druhé číslo: 10
8 10
16 5
False
</pre></td>
<td><pre>
Zadej první číslo: 0
Zadej druhé číslo: 0
0 0
0 0
True
</pre></td>

</tr>
</table>

<h3>3</h3>
<p>Načtěte ze vstupu libovolné číslo (i neceločíselné).<br>Vypočtěte jeho modulo třemi.<br>Umocněte výsledek na druhou.<br>Vydělte výsledek 1/4<br>Spočítejte pátou mocninu výsledku.<br><br>
Vypište pod sebe všechny mezivýsledky.</p>

In [15]:
a = float(input("zadejte první číslo"))
print(a)
a = a % 3
print(a)
a = a**2
print(a)
a = a/0.25
print(a)
a = a**5
print(a)


0.5
0.5
0.25
1.0
1.0


<p>Výsledky pro Vaší kontrolu - referenční řešení má 3 řádky</p>
<table>
<tr>

<td><pre>
Zadej číslo: 1
1.0
1.0
4.0
1024.0
</pre></td>
<td><pre>
Zadej číslo: 3.14
0.14000000000000012
0.019600000000000034
0.07840000000000014
2.9619676669542656e-06
</pre></td>
<td><pre>
Zadej číslo: 0.5
0.5
0.25
1.0
1.0
</pre></td>

</tr>
</table>